solr/core/src/java/org/apache/solr/search/SolrDocumentFetcher.java

```diff
@@ -567,14 +568,62 @@ public void decorateDocValueFields(SolrDocumentBase<?, ?> doc, int docid, Set<St
     final int subIndex = ReaderUtil.subIndex(docid, leafContexts);
     final int localId = docid - leafContexts.get(subIndex).docBase;
     final LeafReader leafReader = leafContexts.get(subIndex).reader();
+    final Set<String> declaredDynamicEventsFields =
+        getDeclaredDynamicEventsFields(localId, leafReader);
     for (String fieldName : fields) {
+      if (declaredDynamicEventsFields != null
+          && fieldName.startsWith("evt_")
+          && !declaredDynamicEventsFields.contains(fieldName)) {
+        continue;
+      }
       Object fieldValue = decodeDVField(localId, leafReader, fieldName);
       if (fieldValue != null) {
         doc.setField(fieldName, fieldValue);
       }
     }
   }
 
+  /**
+   * This is a bit of a hack, to support FullStory-specific dynamicFields of type `evt_*`. As these
+   * fields are user-defined, there may be arbitrarily many of them. Because docValues fields are
+   * column-oriented, Solr must do "pull the column for X field and see if it has a value for this
+   * doc", which can be quite expensive if there is a profusion of matching fields in the index
+   * (even if none of the requested docs contain any of the specified fields).
+   *
+   * <p>Here we pull the specific dynamic field names contained in this doc from a field that is
+   * populated in an UpdateRequestProcessor.
+   */
+  private static Set<String> getDeclaredDynamicEventsFields(int localId, LeafReader leafReader)
+      throws IOException {
+    final SortedSetDocValues dynamicEventsFields =
+        leafReader.getSortedSetDocValues("DynamicEventsFields");
+    final Set<String> hasDynamicEventsFields;
+    if (dynamicEventsFields == null || !dynamicEventsFields.advanceExact(localId)) {
+      return null; // legacy doc, no filtering
+    } else {
+      final int ct = dynamicEventsFields.docValueCount();
+      final BytesRef firstHasFieldName =
+          dynamicEventsFields.lookupOrd(dynamicEventsFields.nextOrd());
+      if (ct == 1) {
+        if (firstHasFieldName.length == 0) {
+          return Collections.emptySet();
+        } else {
+          return Collections.singleton(firstHasFieldName.utf8ToString());
+        }
+      } else {
+        hasDynamicEventsFields = CollectionUtil.newHashSet(ct);
+        assert firstHasFieldName.length != 0;
+        hasDynamicEventsFields.add(firstHasFieldName.utf8ToString());
+        int i = ct - 1;
+        do {
+          hasDynamicEventsFields.add(
+              dynamicEventsFields.lookupOrd(dynamicEventsFields.nextOrd()).utf8ToString());
+        } while (--i > 0);
```

Comment:
is --i this looks good? it will process one less item

Reply:
yes, because we already processed the first one. It's a bit tricky, true. When you get to this part of the code, `ct` will always be `ct > 1`; so in the lowest possible ct (2), we add the first term before the loop, then i is initialized to 1 (`ct - 1`), then we enter the loop, process the second (final) term; then `--i == 0`, so `--i > 0` evaluates to false and we break out of the loop, which is exactly what we want.